### PR TITLE
KAFKA-14834: [7/N] Update VersionedKeyValueStore#put() to return validTo

### PR DESCRIPTION
**Pull request by vcrfxia**
This PR updates the return type of `VersionedKeyValueStore#put(...)` from void to long, where the long is the validTo timestamp of the newly put record, with two special values to indicate either that no such timestamp exists (because the record is the latest for its key) or that the put did not take place (because grace period has elapsed). See [KIP-914](https://cwiki.apache.org/confluence/display/KAFKA/KIP-914%3A+DSL+Processor+Semantics+for+Versioned+Stores) for more.

As part of making this change, VersionedBytesStore introduces its own `put(key, value, timestamp)` method to avoid method signature conflicts with the existing `put(key, value)` method from `KeyValueStore<Bytes, byte[]>` which has void return type. As a result, the previously added `NullableValueAndTimestampSerde` class is no longer needed so it's also been removed in this PR as cleanup.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java [hunk 1]
```diff
@@ -84,22 +86,37 @@
         extends MeteredKeyValueStore<K, ValueAndTimestamp<V>> {
 
         private final VersionedBytesStore inner;
+        private final Serde<V> rawValueSerde;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java:89]: nit: why `raw` -- we usually use `raw` when we have `byte[]` at hand

Comment (vcrfxia) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java:89]: `valueSerde` is already defined in the base class, and I couldn't come up with a better name 😆 Suggestions welcome!

Thinking about it some more, the `valueSerde` defined in the base class is of type `Serde<ValueAndTimestamp<V>>` from the perspective of this class, while this one is `Serde<V>`, so maybe `unwrappedValueSerde` could be a better name? Or is that even more confusing hmm...

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java [hunk 2]
```diff
@@ -84,22 +86,37 @@
...
         }
 
-        @Override
-        public void put(final K key, final ValueAndTimestamp<V> value) {
-            if (value == null) {
-                throw new IllegalStateException("Versioned store requires timestamp associated with all puts, including tombstones/deletes");
+        public boolean put(final K key, final V value, final long timestamp) {
+            Objects.requireNonNull(key, "key cannot be null");
+            try {
+                final boolean isLatest = maybeMeasureLatency(() -> inner.put(keyBytes(key), rawValueSerdes.rawValue(value), timestamp), time, putSensor);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java]: I am confused. I don't see a change to `StreamsMetricsImpl#maybeMeasureLatency` to return a boolean? Can we even use a `Run
nable` as we now need a return type?

Comment (vcrfxia) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java]: There's another existing implementation which takes `Supplier<T>` and passes the result through: https://github.com/apache/kafka/blob/f252c75bf3f2db93bf062bc7e49d3c79a06e9568/streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java#L876

It's the same one that's currently being used in the file to return `ValueAndTimestamp<V>` from `get(key)` in the method below this one.

File: streams/src/main/java/org/apache/kafka/streams/state/VersionedKeyValueStore.java [hunk 3]
```diff
@@ -41,6 +42,9 @@
  */
 public interface VersionedKeyValueStore<K, V> extends StateStore {
 
+    long PUT_RETURN_CODE_VALID_TO_UNDEFINED = -1L;
+    long PUT_RETURN_CODE_NOT_PUT = Long.MIN_VALUE;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/VersionedKeyValueStore.java:46]: Should both be `final` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java:89]: Maybe `plainValueSerde` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java]: Ah. Missed this one. Nice. Was really wondering as your PR compiled :)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java [hunk 4]
```diff
@@ -641,6 +642,8 @@ private <T extends VersionedStoreSegment> PutStatus maybePutToLatestValueStore(
                     versionedStoreClient.deleteLatestValue(key);
                 }
                 return new PutStatus(true, foundTs);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java:644]: Seems `foundTs` should always be `SENTINEL_TIMESTAMP` here? Should we pass in `SENTINEL_TIMESTAMP` directly to make it more explicit?

For this case, we might also not need to introduce `latestValueStoreTimestamp` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java [hunk 5]
```diff
@@ -825,7 +835,7 @@ private <T extends VersionedStoreSegment> void finishPut(
                     if (RocksDBVersionedStoreSegmentValueFormatter.getNextTimestamp(rawSegmentValue) == timestamp) {
                         // next timestamp equal to put() timestamp already represents a tombstone,
                         // so no additional insertion is needed in this case
-                        return;
+                        return foundTs;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java:838]: Same: should we return `SENTINEL_TIMESTAMP` directly?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java [hunk 6]
```diff
@@ -837,6 +847,7 @@ private <T extends VersionedStoreSegment> void finishPut(
                     segment.put(key, segmentValue.serialize());
                 }
             }
+            return foundTs;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java:850]: Again: should we return `SENTINEL_TIMESTAMP` directly?

We could also skip the `return` here (and at the end of `else` block) and have a final `return foundTs`  instead?

**Pull request merged**